src/Stratis.Bitcoin/P2P/Protocol/Behaviors/EnforcePeerVersionCheckBehavior.cs

```diff
@@ -0,0 +1,105 @@
+
+        /// <summary>Logger factory used while cloning the object.</summary>
+        protected readonly ILoggerFactory LoggerFactory;
+
+        /// <summary>Instance logger.</summary>
+        protected readonly ILogger Logger;
+
+        /// <summary>
+        /// Set to <c>true</c> if the attached peer callbacks have been registered and they should be unregistered,
+        /// <c>false</c> if the callbacks are not registered.
+        /// </summary>
+        protected bool CallbacksRegistered;
+
+        /// <summary>
+        /// Initializes an instance of the object for outbound network peers.
+        /// </summary>
+        /// <param name="chainIndexer">The chain of blocks.</param>
+        /// <param name="nodeSettings">User defined node settings.</param>
+        /// <param name="network">Specification of the network the node runs on - regtest/testnet/mainnet.</param>
+        /// <param name="loggerFactory">Factory for creating loggers.</param>
+        public EnforcePeerVersionCheckBehavior(ChainIndexer chainIndexer,
+            NodeSettings nodeSettings,
+            Network network,
+            ILoggerFactory loggerFactory)
+        {
+            this.ChainIndexer = chainIndexer;
+            this.NodeSettings = nodeSettings;
+            this.Network = network;
+            this.LoggerFactory = loggerFactory;
+            this.Logger = loggerFactory.CreateLogger(this.GetType().FullName, $"[{this.GetHashCode():x}] ");
+        }
+
+        [NoTrace]
+        protected Task OnMessageReceivedAsync(INetworkPeer peer, IncomingMessage message)
+        {
+            int enforceMinProtocolVersionAtBlockHeight = this.Network.Consensus.Options.EnforceMinProtocolVersionAtBlockHeight;
+            bool enforcementEnabled = enforceMinProtocolVersionAtBlockHeight > 0;
+            bool enforcementApplied = this.NodeSettings.MinProtocolVersion >= this.Network.Consensus.Options.EnforcedMinProtocolVersion;
+
+            bool enforcementHeightReached = this.ChainIndexer.Height >= enforceMinProtocolVersionAtBlockHeight;
+            if (enforcementEnabled && !enforcementApplied && enforcementHeightReached)
+            {
+                this.Logger.LogDebug("Changing the minumum supported protocol version from {0} to {1}.", this.NodeSettings.MinProtocolVersion, this.Network.Consensus.Options.EnforcedMinProtocolVersion);
+                this.NodeSettings.MinProtocolVersion = this.Network.Consensus.Options.EnforcedMinProtocolVersion;
+            }
+
+            // The statement below will close connections in case the this.NodeSettings.MinProtocolVersion has changed during node execution.
+            if (peer?.PeerVersion?.Version != null && peer.PeerVersion.Version < this.NodeSettings.MinProtocolVersion)
+            {
+                this.Logger.LogError("Unsupported client version, dropping connection.");
+                this.AttachedPeer.Disconnect("Peer is using unsupported client version");
```

Comment:
As discussed `this.AttachedPeer` might become null you should take a scope variable of `AttachedPeer` at the start of the method.

Reply:
That check is not required as can be seen in other behaviour: https://github.com/stratisproject/StratisBitcoinFullNode/blob/d7e57352e01fdbaf325c3c53223bb43a93cc5417/src/Stratis.Bitcoin.Features.Consensus/Behaviors/ProvenHeadersReservedSlotsBehavior.cs#L120

Reply:
yes but register is called inside attached core which is "alsmot" guaranteed to not have the peer as null (but I would say its wrong also there) in the method MessageReceived it's quite likely you'll hit that AttachedPeer is null (at some point)

We cannot control when AttachedPeer is null it can happen at any point in time (a peer can issue a diconnect or the node may disconnect for what ever reason)